### PR TITLE
Reorganize Test Folder structure

### LINKED ISSUE #297: How to seperate Unit and Integration tests
We are currently in a state where we get some integration tests. therefore we should investigate a solution to seperate them from regular unit tests.

Once we found a solution, i will add a ADR when realizing the chosen option

Possible Approaches:
- Use JUnit [Categories](https://github.com/junit-team/junit4/wiki/categories)
	- `+` Supported by Maven and IntelliJ (Categories to execute can be selected)
	- `+` Quite easy to use/setup
	- `-` Both test categories feature the same `source` and `resources` folder
- Use Seperated `integration` and `test` folders [See here](https://www.petrikainulainen.net/programming/maven/integration-testing-with-maven/)
	- `+` Clear differentiation between integration and unit tests
	- `-` Integration tests have to contain a specific string (e.g. the files have to end with `IT.java`)
- Put the integration tests in a seperate module
	- `+` Clear differentiation between integration and unit tests (even clearer than with a IT folder)
	- `-` Possible problems when computing the coverage from integration tests

I personally would prefer the first option (JUnit Categories). Because it is still possible to seperate the tests by putting them in different packages. (As long as there is no issue with visibilies)

### PR DESCRIPTION
**Pull request by c-mueller**
Once done this PR resolves #297.

The following tasks still have to be done:
- [x] Add ADR for the chosen folder Structure
- [x] Create a small note on how to use the modified environment
- [x] Check if the integration resource folder is loaded
- [x] Collect Coverage Data from integration tests
- [x] Provide default intelij execution Profiles
- [x] Update Travis.yml
- [x] Update Jenkinsfile

Merging this results in the following consequence:
You have to add the Resource Directory of the integration tests manually using:
unbenannt

Right Click the Directory -> "Mark Directory as" -> "Test Resource Root"

Please also note that the integration folder and the test folder still result in the same classpath when executing. Class Names and paths in the resource folder have to be unique when combining the two directories.

### CODE REVIEW

File: docs/adr/0011-put-integration-tests-in-seperate-folder.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+* Use JUnit Categories and the same `test` directory to seperate between Integration and Unit Tests
+* Store Integration and Unit Tests in seperate directories and use JUnit Categories to categorize them
+* Store Integration and Unit Tests in seperate directories and use a required Suffix on Integration Test Classes
+* Migrate project to gradle
+
+## Decision Outcome
+
+* Chosen Alternative: Store Integration and Unit Tests in seperate directories and use JUnit Categories to categorize them
+* While still Clearly seperating Unit and Integration tests this method still allows a fine grained categorisation of tests.
+* Consequences: The resource directory of the `ìntegration` folder has to be manually set within the IDE
```
Comment (hnicke) [inline docs/adr/0011-put-integration-tests-in-seperate-folder.md]: I think this is not a consequence: I didn't have to do it, was done automatically

File: docs/dev/server-testing.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# Testing Guidelines for the Server module
```
Comment (hnicke) [inline docs/dev/server-testing.md]: Please correct spelling and grammar mistakes in this file. I've found a handful.

File: server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package org.opentosca.toscana.core;
+
+import org.opentosca.toscana.IntegrationTest;
+
+import org.junit.experimental.categories.Category;
+
+@Category(IntegrationTest.class)
+public abstract class BaseJUnitIntegrationTest extends BaseJUnitTest {
```
Comment (hnicke) [inline server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java]: What the heck is a BaseJUnitIntegrationTest? :grin:
A test is either an integration or an unit test; never a UnitIntegrationTest. I am confused. Can you explain?

File: server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/image/DockerImageBuilderIT.java [hunk 4]
```diff
@@ -44,12 +47,12 @@ public void testBuildShaImage() throws Exception {
     }
 
     private boolean isDockerAvailable() {
```
Comment (hnicke) [inline server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/image/DockerImageBuilderIT.java]: Do we still need this at all?

File: server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTagLoaderIT.java [hunk 5]
```diff
@@ -19,10 +21,11 @@
 import static org.opentosca.toscana.model.capability.OsCapability.builder;
 
 @ActiveProfiles( {"bimloader"})
```
Comment (hnicke) [inline server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTagLoaderIT.java]: Do we still need the profile?

File: server/src/test/java/org/opentosca/toscana/UnitTest.java [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package org.opentosca.toscana;
+
+public interface UnitTest {
```
Comment (hnicke) [inline server/src/test/java/org/opentosca/toscana/UnitTest.java]: Why is this required?

File: server/src/integration/java/org/opentosca/toscana/IntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package org.opentosca.toscana;
+
+public interface IntegrationTest {
```
Comment (hnicke) [inline server/src/integration/java/org/opentosca/toscana/IntegrationTest.java]: Why is this required?

Comment (c-mueller) [reply server/src/integration/java/org/opentosca/toscana/IntegrationTest.java]: This interface and the UnitTest one represents a [Test category in JUnit](https://github.com/junit-team/junit4/wiki/categories)

File: server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java (hunk 3 above)
Comment (c-mueller) [reply server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java]: Well it represents a Integration test that does not need a spring context. Didn't think of a better name. But maybe `BaseIntegrationTest` is better

File: server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/image/DockerImageBuilderIT.java (hunk 4 above)
Comment (c-mueller) [reply server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/image/DockerImageBuilderIT.java]: I would keep it. just to be save. When trying to run the integration tests with no docker daemon available this test will fail.

File: server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTagLoaderIT.java (hunk 5 above)
Comment (c-mueller) [reply server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTagLoaderIT.java]: Yes.

File: server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java (hunk 3 above)
Comment (hnicke) [reply server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java]: Yes! Maybe also rename BaseJunitTest to BaseUnitTest.

File: docs/adr/0011-put-integration-tests-in-seperate-folder.md (hunk 1 above)
Comment (c-mueller) [reply docs/adr/0011-put-integration-tests-in-seperate-folder.md]: Lets say it might cause issues. Because as I said already i had to set it manually even after restarting the IDE

File: server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java (hunk 3 above)
Comment (hnicke) [reply server/src/integration/java/org/opentosca/toscana/core/BaseJUnitIntegrationTest.java]: Yeah technically BaseJUnitIntegrationTest is not a contradiction; but then `JUnit` information is quite unnecessary..

**Comment by hnicke** [comment]
Good work!
